### DISCUSSION TITLE
Part of NextJS app no longer working - ChunkLoadError (critical in prod)

### DISCUSSION THREAD

**Discussion by AlexanderMoonBit** [vercel/next.js#48328] [Help] [answered]
### Summary

We are recently seeing errors in the Production and even localhost (npm run build/run start) where on some pages it gives "Application error: a client-side exception has occurred (see the browser console for more information)". When I check the console it shows the following error

ChunkLoadError: Loading chunk 5760 failed.
(error: http://localhost:3000/_next/static/chunks/5760-40c839657ea31a15.js)
    at d.f.j (webpack-a66c4bcffce5144a.js:1:3328)
    at webpack-a66c4bcffce5144a.js:1:1385
    at Array.reduce (<anonymous>)
    at d.e (webpack-a66c4bcffce5144a.js:1:1351)
    at self.__next_chunk_load__ (5782-03768f85683b0a9b.js:1:5780)
    at 5782-03768f85683b0a9b.js:9:2777
    at 5782-03768f85683b0a9b.js:9:3187
    at M (5782-03768f85683b0a9b.js:9:3390)
    at t (5782-03768f85683b0a9b.js:9:5961)

<img width="1850" alt="image" src="https://user-images.githubusercontent.com/103491704/231690984-566ac74d-b418-40c9-a563-183a2755ed1d.png">

We tried different things like
- Change the nextjs version
- remove .next and rerun npm run build

our next.config.json

```
// next.config.js
const nextConfig = {
  experimental: {
    appDir: true,
  },
  modularizeImports: {
    '@react-feather': {
      transform: '@react-feather/{{member}}',
    },
  },
  images: {
    remotePatterns: [
      {
        protocol: 'https',
        hostname: 'private', // development domain
        port: '',
        pathname: '/storage/**',
      },
      {
        protocol: 'https',
        hostname: 'private', // staging domain
        port: '',
        pathname: '/storage/**',
      },
      {
        protocol: 'https',
        hostname: 'private', // production domain
        port: '',
        pathname: '/storage/**',
      },
    ],
  },
  async rewrites() {
    return [
      {
        source: '/private/:gcloud_url/:path*',
        destination:
          'https://:gcloud_url/:path*',
      },
    ];
  },
};

module.exports = nextConfig;
```

Our package.json

```
{
  "name": "moonbit-web-application",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@lottiefiles/lottie-player": "^1.7.1",
    "@next/font": "13.1.6",
    "@passbase/button": "^4.4.0",
    "@sumsub/websdk": "^1.3.7",
    "@sumsub/websdk-react": "^1.3.7",
    "@supabase/auth-helpers-nextjs": "^0.5.5",
    "@supabase/supabase-js": "^2.8.0",
    "@types/node": "18.14.0",
    "@types/react": "18.0.28",
    "@types/react-dom": "18.0.11",
    "chart.js": "^4.2.1",
    "chartjs-adapter-date-fns": "^3.0.0",
    "encoding": "^0.1.13",
    "eslint": "8.34.0",
    "eslint-config-next": "13.1.6",
    "flowbite": "^1.6.3",
    "next": "^13.2.5-canary.20",
    "qrcode.react": "^3.1.0",
    "react": "18.2.0",
    "react-chartjs-2": "^5.2.0",
    "react-copy-to-clipboard": "^5.1.0",
    "react-dom": "18.2.0",
    "react-feather": "^2.0.10",
    "react-icons": "^4.7.1",
    "server-only": "^0.0.1",
    "sharp": "^0.32.0",
    "swr": "^2.1.0",
    "typescript": "4.9.5"
  },
  "devDependencies": {
    "@types/react-copy-to-clipboard": "^5.0.4",
    "autoprefixer": "^10.4.13",
    "postcss": "^8.4.21",
    "tailwindcss": "^3.2.7"
  }
}
```

Help appreciated!

**Comment by icyJoseph** [comment]
So the chunks is not being found on the server, which gives a 404 and from there the client falls apart. 

The question is why would it fail to get, or find, the chunk. Are you using middleware, or some other network layer that could be making the request fail?

**Reply by AlexanderMoonBit** [reply]
On the page affected, there is nothing really happening - no middleware or requests. We made some page restructuring on that section but can't see why would still cache after removing .next 

We have this page /deposit that throws the error. Also /deposit/crypto/details throw the error but not /deposit/crypto


```
import Select from "./components/select"


export default function Deposit() {
    return(
        <>
        <span className="text-primary font-semibold text-3xl">Select deposit method</span>
        <Select/>
        </>
    )
}

```

And select

```
import Link from "next/link";
import Image from "next/image";

export default function Select () {
    return (
        <div className="flex flex-col lg:flex-row gap-8 mt-14">
          <Link 
          href="/deposit/crypto"
          className="flex flex-col border-2 border-primary bg-primary rounded-3xl p-8 gap-4 flex-grow h-80 lg:w-1/2 hover:transform hover:-translate-y-1 transition-transform duration-300"
          > 
            <div className="flex items-center justify-center gap-4 mb-4">
            <Image
                    src="/btc.png"
                    width={56}
                    height={56}
                    alt="Bitcoin logo"
            />
                <h2 className="text-2xl text-center text-white font-semibold">Crypto</h2>
                
            </div>
              <p className="text-white font-regular">Deposit cryptocurrency directly into your Moonbit wallet. This is the cheapest method.</p>


          </Link>
          <Link href="/deposit/fiat"
          className="flex flex-col border-2 border-primary rounded-3xl p-8 gap-4 h-80 flex-grow lg:w-1/2 hover:transform hover:-translate-y-1 transition-transform duration-300"
          >
              <div className="flex items-center justify-center gap-4 mb-4">
              <Image
                    src="/transak.png"
                    width={56}
                    height={56}
                    alt="Transak logo"
            />
                <h2 className="text-2xl text-center text-primary font-semibold">Fiat</h2>
                
              </div>
              <p className="text-primary font-regular">Buy cryptocurrency with your local currency using our partner Transak. With Transak, you can easily purchase Bitcoin, Ethereum, and other popular cryptocurrencies. This is the most expensive method.</p>
          </Link>
        </div>
      );
}
```

**Reply by icyJoseph** [reply]
- You don't have any service worker installed right?
- How's your application deployed?

**Reply by AlexanderMoonBit** [reply]
@icyJoseph 
- No service workers 
- We use Github with Actions to Cloud Run.

I checked out a one-day-old commit or so and that one works but it has an old structure inside the folder app/deposit

In that commit deposit/details is not working for a similar Chunk error reason - but it used to work

**Reply by AlexanderMoonBit** [reply]
@icyJoseph okay I found the culprit it seems. It's related to nextJS component Image in the shared code above. By using img instead of Image from nextjs it doesn't throw error - I have many places where I use Image but it doesn't cause an issue, sooo random...

**Reply by icyJoseph** [reply]
Oh wow, really? What next version do you have? There must be something on that path that's making it fail. Is it all happening in appDir?

**Reply by AlexanderMoonBit** [reply]
using the next version 13.3 (latest) and yes in appDir

**Reply by icyJoseph** [reply]
I see you were using canary, does it happen downgrading to next@latest or something before 13.3?

**Comment by AlexanderMoonBit** [comment]
I tried a different version (13.2) but it didn’t work. I tried canary since
I saw a mention about a fix that I thought may help.

On Thu, 13 Apr 2023 at 12:06, Joseph ***@***.***> wrote:

> I see you were using canary, does it happen downgrading to ***@***.*** or
> something before 13.3?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vercel/next.js/discussions/48328#discussioncomment-5603044>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYVSQ6ESB4EW2FC6MVLFHU3XA7M2ZANCNFSM6AAAAAAW4WHTYM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Alexander Thomsen

Founder

Moonbit
***@***.***
www.moonbit.ai

**Reply by NhatLam310399** [reply]
I had same error. How do u fixed ? thanks

**Reply by AlexanderMoonBit** [reply]
It was nextjs Image component - changed to img and it fixed it

**Reply by anil2612coder** [reply]
it works

**Reply by shiv-anya** [reply]
Then are we not letting the whole purpose of next image fail? Which is cache images for high performance. This is a temporary solution with bad performance in return.

**Reply by taufikmuharrom** [reply]
> It was nextjs Image component - changed to img and it fixed it

I just tried this solution, and it works! Thanks, @AlexanderMoonBit

**Comment by AlexanderMoonBit** [comment]
It was nextjs Image component - changed to img and it fixed it

On Sat, 13 May 2023 at 02:26, Nhật Lâm ***@***.***> wrote:

> I had same error. How do u fixed ? thanks
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vercel/next.js/discussions/48328#discussioncomment-5889705>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYVSQ6A6COECVS3ESGGKEJLXF3PNRANCNFSM6AAAAAAW4WHTYM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Alexander Thomsen

Founder

Moonbit
***@***.***
www.moonbit.ai

**Comment by mattdev-sh** [comment]
I had same error after builded my next app with export config active, to have a static a website and deploy it on apache. With mamp in localhost, there's no problems any but on prod yes.

If problem came out on nginx prod also, i don't understand why the problem is not come out in more topics.  

For resolve on apache prod i had to change manually path [locale ] ( my [lang] folder ) with simple "locale" in all files in /out folder and renamed folder also with the same name. 

This problem seems happen because a codification of chrome about the name of folder [locale]. I will try on nginx making a not static build, but for now, for who have this problems on apache, this is a manually solutions.

**Comment by tlchatt** [comment]
Also having this problem, 
Site serves the chunks via IP but not via the domain. 
Only fails on the page chunk
![image](https://github.com/vercel/next.js/assets/26937732/8624f248-b60a-4937-9f91-c8e3c86d64fc)
First set of fetches to _next work (via the ip)
Second set mostly work minus the page chunk (via the https://domain)
The chunk is appearing in a dyamic page function console log as slug params, app/[..slug]/page.js

If I turn dynamicParams off it stops showing but still fails to load the chunk.
export const dynamicParams = false // true | false,

**Comment by sudeepgumaste** [comment]
Leaving it here so that it might help someone. In my case it was just happening on production and not on local, I was getting a 301 for when the /path/[chunk] was being requested. This ended happening because of a redirect rule I had on cloudflare that was redirecting any /path to a different location. So when next tried fetching that particular chunk, the redirect rule came into play and redirecting the request to a completely different url where obviously it wasn't able to get the chunk. Turning off the redirect rule fixed it

**Comment by benjick** [comment]
Over a year and no real answer?

**Comment by NevilDabre** [comment]
To achieve custom file naming and handle various other requirements, you can extend the Webpack configuration in your `next.config.ts`. Below is a breakdown of what the Webpack configuration does and how it helps control the output of both JavaScript and CSS files:

1. Custom JavaScript File Naming
The configuration ensures that JavaScript files are placed in static/chunks/ and named according to their chunk name. This helps maintain consistency and organization of bundled files:

```javascript
config.output.filename = () => `static/chunks/[name].js`;
config.output.chunkFilename = () => `static/chunks/[name].js`;
``` 

2. Handling SCSS to CSS Conversion and Naming
Using the MiniCssExtractPlugin, the configuration extracts and names your CSS files (generated from SCSS) into the static/css/ folder with the same naming convention ([name].css):

```javascript
new MiniCssExtractPlugin({
  filename: () => `static/css/[name].css`,
  chunkFilename: () => `static/css/[name].css`,
});
``` 

Full Webpack code `next.config.ts`
```javascript
  webpack: (config, { isServer, buildId, dev }) => {
    // Fixes npm packages that depend on `fs` module
    if (!isServer) {
      config.resolve.fallback.fs = false;
    }
    /**
     * Returns environment variables as an object
     */
    const env = Object.keys(process.env)
      .filter((key) => !key.includes('NEXT_'))
      .reduce((acc, curr) => {
        acc[`process.env.${curr}`] = JSON.stringify(process.env[curr]);
        return acc;
      }, {});

    /** Allows you to create global constants which can be configured
     * at compile time, which in our case is our environment variables
     */
    config.plugins.push(new webpack.DefinePlugin(env));

    //Required to prevent "React is not defined" error
    config.plugins.push(new webpack.ProvidePlugin({
      React: "react",
    }))

    if (!isServer && !dev) {
      config.plugins.push(new CustomChunkNamingPlugin());

      config.output.filename = () => {
        return `static/chunks/[name].js`;
      };

      config.output.chunkFilename = () => {
        return `static/chunks/[name].js`;
      }
    }

    if (!dev) {
      config.plugins.push(
        new MiniCssExtractPlugin({
          filename: () => {
            return `static/css/[name].css`;
          },
          chunkFilename: () => {
            return `static/css/[name].css`;
          },
        })
      );

      // Ensure the splitChunks configuration exists
      //if (!config.optimization.splitChunks) {
        config.optimization.splitChunks = {
          chunks: 'all',
          maxSize: 240000,
        };
      //}

      // Adjust the splitChunks configuration for styles
      config.optimization.splitChunks.cacheGroups = {
        ...config.optimization.splitChunks.cacheGroups,
        vendor: {
          test: /[\\/]node_modules[\\/]/,
          name: 'vendors',
          maxSize: 240000,
          chunks: 'all',
        },
        styles: {
          name: false,
          test: /\.(css|scss)$/,
          chunks: 'all',
          enforce: true,
        },
      };

      // Add rules for processing CSS and SCSS files
      config.module.rules.push(
        {
          test: /\.module\.(css|scss)$/,
          use: [
            MiniCssExtractPlugin.loader,
            {
              loader: 'css-loader',
              options: {
                modules: {
                  localIdentName: '[name]_[local]__[hash:base64:5]',
                },
                importLoaders: 2,
              },
            },
            {
              loader: 'resolve-url-loader',
              options: {
                sourceMap: dev,
              },
            },
            'sass-loader',
          ],
        },
        {
          test: /\.(css|scss)$/,
          exclude: /\.module\.(css|scss)$/,
          use: [
            MiniCssExtractPlugin.loader,
            'css-loader',
            {
              loader: 'resolve-url-loader',
              options: {
                sourceMap: dev,
              },
            },
            'sass-loader',
          ],
        }
      );
    }

    return config;
  },
``` 

This setup gives you full control over file naming, CSS/SCSS handling, and performance optimizations in your Next.js project.

**Reply by abhibhaw** [reply]
Is this working?

**Comment by NJneeraj** [comment]
Same error.Any solutions?

**Comment by abhibhaw** [comment]
Any working solution? Seems like need to try something else and not next

**Comment by kassiwin** [comment]
We’re encountering this issue in production where users on Chrome 83 and Safari 12 (and below) are unable to access the application. This is causing significant frustration for both our users and our team.

We’re currently using Next.js 14.2.24, and this issue seems to be directly impacting user experience for those on older browser versions.

<img width="942" alt="image" src="https://github.com/user-attachments/assets/f9809120-5d33-48d2-bf06-93562bf147c7" />

Any solutions of suggestion?

**Reply by NevilDabre** [reply]
If your server is behind Cloudflare or another CDN, check whether caching policy is causing the issue. The error occurred because Cloudflare cached an outdated file with a mismatched path, leading to a 404 error. To fix this, ensure consistent file naming across deployments to prevent chunk errors. You can refer this explanation - [https://github.com/vercel/next.js/discussions/48328#discussioncomment-10894711](https://github.com/vercel/next.js/discussions/48328#discussioncomment-10894711)

**Reply by santosguilherme** [reply]
@kassiwin I think the `ChunkLoadError`s are happening because of the `SyntaxError`s. The browser cannot parse the chunk file, usually it happens in older browsers. You can fix your target / browserlist options or use the browserlist + transpile modules from Nextjs if it's a dependency,

**Comment by ddfonseca** [comment]
Any working solution? It seems like everyone is trying different things

**Comment by egold** [comment]
In case anyone is still experiencing this: here's why I was experiencing it, and how I fixed it:

1. Why I was experiencing it: Running next in a Docker container as a standalone server (`output: "standalone"` in `next.config.js`)
2. How I fixed it: [As explained here](https://nextjs.org/docs/pages/api-reference/config/next-config-js/output), Next assumes if you're running a standalone server, that you're hosting your chunks in some CDN somewhere. So anything in your `public/` and `static/` directories are not included in the output. You need to be sure to include them. Here's the snippet at time of writing:

```
cp -r public .next/standalone/ && cp -r .next/static .next/standalone/.next/
```

Or, in a Dockerfile:

```
RUN cp -r public .next/standalone/ && cp -r .next/static .next/standalone/.next/
```

It seems like a pretty opinionated and large assumption by Next that just because you're Dockerizing your app, you are default going to host your chunks on CDNs. Some of us just want to Dockerize for other reasons, such as consistent dev environments, running on Docker-based PaaS like Aptible, etc. And at the same time, once an app reaches the level of sophistication where static assets are being hosted on a CDN, it's probably a more mature app with a DevOps person or two in the picture. It seems like including static assets in standalone mode should be the default. 🤷

**Comment by webrdaniel** [comment]
In our instance, the issue was related to browser caching. During a new deployment, existing files were still cached and referencing invalid chunks, resulting in seemingly random crashes. This was caused by the `stale-while-revalidate` header, which was by default set to a very high value. We successfully resolved this by removing the header by setting the [expireTime](https://nextjs.org/docs/app/api-reference/config/next-config-js/expireTime) in the `next.config.js` file to zero. `expireTime: 0`

If that does not help. You might also explore passing the [deployment ID](https://nextjs.org/docs/app/guides/self-hosting#version-skew) that will ensure unique chunk names between deployments.

**Reply by reedwane** [reply]
Thanks a lot for writing this. I tried the deploymentId and it works fine.
Here's the new link to the doc for that though: https://nextjs.org/docs/app/guides/self-hosting#build-cache

**Comment by bgw** [comment, marked as answer]
I've left a detailed explanation here about how `ChunkLoadError`s can happen, and how to deal with them here: https://github.com/vercel/next.js/discussions/82651#discussioncomment-16607401

While these issues are impossible to completely avoid (it can be an issue with the user's network connection), **we recommend enabling skew protection**, we changed logic for Turbopack in Next.js 16.1 to better match Sentry's built-in filters, and we're shipping chunk load retry logic with Turbopack in 16.3 to reduce the frequency of these errors. If you're using webpack, you can try using the third-party `webpack-retry-chunk-load-plugin` to mitigate some portion of these errors.
